pl-faded-parsons.py

```diff
@@ -1,314 +1,538 @@
+    `out_filename` : `str`
+        The file to which to include the student's submission. Specified with `file-name="..."`.
+    `size` : `Literal["narrow", "wide"]`
+        The size of the solution tray. `"narrow"` indicates it should take approximately half the width of the problem pane. `"wide"` indicates it should take the full width of the problem pane.
+    `solution_path` : `str`
+        The path to a file containing the solution. Specified with `solution-path="..."`. Raises `FileNotFoundError` on access if not found.
+    `solution` : `str`
+        The solution. Specified with `solution-path="..."`. Raises `FileNotFoundError` on access if file not found.
+    `trays` : `Submission.Trays`
+        The trays used in this problem. MUST CALL `.load(...)` TO DEFINE.
+    `log` : `List[Submission.LogEntry]`
+        The log of events for this problem. MUST CALL `.load(...)` TO DEFINE.
+
+    Methods
+    -------
+    `to_mustache() -> Mustache`
+        Produce a `Mustache` instance for rendering
+    `to_code() -> str`
+        Compile the student submission into an executable code snippet.
+
+    """
+
+    class Formats(Enum):
+        BOTTOM = "bottom"
+        RIGHT = "right"
+        NO_CODE = "no_code"
+
+    @property
+    def solution_path(self) -> str:
+        if not os.path.exists(self._solution_path):
+            raise FileNotFoundError('\n'
+                f'\tCorrect answer not found at `{self._solution_path}`! \n'
+                 '\tProvide an answer or set "showCorrectAnswer" to false in `./info.json`'
+            )
+
+        return self._solution_path
+
+    @property
+    def solution(self) -> str:
+        with open(self.solution_path, "r") as f:
+            return f.read()
+
+    def __init__(self, element_html: str, data: pl.QuestionData):
+
+        element: xml.HtmlElement = xml.fragment_fromstring(element_html)
+        self._element: xml.HtmlElement = element
+        pl.check_attribs(
+            element,
+            required_attribs=[
+                "answers-name",
+            ],
+            optional_attribs=[
+                "solution-path",
+                "format",
+                "language",
+                "file-name",
+                "solution-path"
+            ]
+        )
+
+        self.answers_name: str = pl.get_string_attrib(element, 'answers-name', '')
+        self.format = FadedParsonsProblem.Formats(
+            pl.get_string_attrib(element, "format", "right")
+                .replace("-", '_')
+        )
+        self.pre_text  = get_child_text_by_tag(element, "pre-text").strip("\n")
+        self.post_text = get_child_text_by_tag(element, "post-text").strip("\n")
+        self.language: str = pl.get_string_attrib(element, "language", "")
+        self.out_filename = pl.get_string_attrib(element, 'file-name', 'user_code.py')
+        self.size = "narrow" if self.format == FadedParsonsProblem.Formats.RIGHT else "wide"
```

Comment:
Bug here: self.format appears to be a string, but to do comparison to Enum variants, use `Formats.RIGHT.value`. Without `.value`, is false. This may not hold for StrEnum, may be worth checking...

Reply:
Maybe the better solution is to make `self.format: FadedParsonsProblem.Formats`

Reply:
Last note, enums are always singular.

Reply:
~nvm~